### PR TITLE
documentation: S3C-2675 add audit logs requirements document

### PR DESCRIPTION
**Pull request by terrydervaux**
Add audit logs requirements document for S3 components.

### CODE REVIEW

File: docs/requirements/AUDIT-LOGS.md [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+* In case of an integration with an External Identity Provider(using web identity or SAML federation), the 'who' must be the External Identity login or username
+* Failed action must appear in the Audit logs with the reason of failed (e.g: wrong signature, or privileges...)
+* Audit logs should be included into sreport
+* Audit logs rotation, purge, activation, desactivation must be configurable
+* Audit logs configuration, activation, desactivation must be documented
+* Audit logs configuration, activation, desactivation must persist to Scality upgrade
+
+### Who definition
+
+'Who' can be:
```
Comment (RachedBenMustapha) [inline docs/requirements/AUDIT-LOGS.md:26]: There is the effective identity that performs an action, and the actual identity behind. For example in the case of a scheduled task, the "who" can be a service account, on its own, but for an asynchronous task submitted by a real user and performed by a service account, the "who" is the combination of:
- the real user that initiated the action
- the service account that performed the action on behalf of the user above
- the principal used (a user can have several ways to identify, like several access keys, roles to assume, etc)

Here service account is very broadly defined, can be unix-level or IAM level or database-level.

File: docs/requirements/AUDIT-LOGS.md [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+
+## Context
+
+Some customers are using tools like [IBM QRadar](https://www.ibm.com/fr-fr/security/security-intelligence/qradar) in order to centralize and monitor who did what on IT systems for security reason. This tool is ingesting events formatted in Leefx format, which is a basic key/value list, separated with a space.
+
+## Requirements
+
+### Storie
+
+As a Storage Administrator, I want audit logs that describe who perform an action on a S3 or IAM ressource, in order connect my Scality products to my IBM QRadar system.
```
Comment (RachedBenMustapha) [inline docs/requirements/AUDIT-LOGS.md:11]: Focusing only on S3/IAM level events/actions is an incomplete way of tracking changes to the system: actions that relate to security (and can open breaches, unauthorized access, broken functionality) can be in other layers such as system/unix, database, orchestrator, authentication subsystem.

If we want to exclude these other layers, I suggest stating it explicitly in the specification.

File: docs/requirements/AUDIT-LOGS.md [hunk 3]
```diff
@@ -0,0 +1,292 @@
...
+
+* Audit logs must be formatted using LEEF format
+* Audit logs must be self-descriptive, that means a Storage Administrator don't need to correlate multiple log lines or log files to know who perform an action on a ressource
+* In case of an integration with an External Identity Provider(using web identity or SAML federation), the 'who' must be the External Identity login or username
+* Failed action must appear in the Audit logs with the reason of failed (e.g: wrong signature, or privileges...)
+* Audit logs should be included into sreport
+* Audit logs rotation, purge, activation, desactivation must be configurable
+* Audit logs configuration, activation, desactivation must be documented
+* Audit logs configuration, activation, desactivation must persist to Scality upgrade
+
```
Comment (RachedBenMustapha) [inline docs/requirements/AUDIT-LOGS.md:23]: There are durability, immutability, resilience and performance considerations as well.

File: docs/requirements/AUDIT-LOGS.md (hunk 2 above)
Comment (nootal) [reply docs/requirements/AUDIT-LOGS.md:11]: Agreed. It's only related to S3 components Scality is delivering. We don't care about logging the SSH connections here.

File: docs/requirements/AUDIT-LOGS.md [hunk 4]
```diff
@@ -0,0 +1,292 @@
...
+## Requirements
+
+### Storie
+
+As a Storage Administrator, I want audit logs that describe who perform an action on a S3 or IAM ressource, in order connect my Scality products to my IBM QRadar system.
+
+Acceptance Criteria:
+
+* Audit logs must be formatted using LEEF format
+* Audit logs must be self-descriptive, that means a Storage Administrator don't need to correlate multiple log lines or log files to know who perform an action on a ressource
```
Comment (nootal) [inline docs/requirements/AUDIT-LOGS.md:16]: Careful here, because the `who` can be misleading. If you only have a key to access a resource, you may need to correlate with other sources to have a name for example.

File: docs/requirements/AUDIT-LOGS.md [hunk 5]
```diff
@@ -0,0 +1,292 @@
+# [S3C-2675](https://scality.atlassian.net/browse/S3C-2675): Audit logs requirements document
+
+## Context
+
+Some customers are using tools like [IBM QRadar](https://www.ibm.com/fr-fr/security/security-intelligence/qradar) in order to centralize and monitor who did what on IT systems for security reason. This tool is ingesting events formatted in Leefx format, which is a basic key/value list, separated with a space.
```
Comment (anurag4DSB) [inline docs/requirements/AUDIT-LOGS.md:5]: It would be nice to know which LEEF version are we going to transform out logs to.
The headers look something like this:
• LEEF:Version|Vendor|Product|Version|EventID|
• LEEF:1.0|Microsoft|MSExchange|4.0 SP1|15345|
• LEEF:2.0|Lancope|StealthWatch|1.0|41|^|

File: docs/requirements/AUDIT-LOGS.md [hunk 6]
```diff
@@ -0,0 +1,292 @@
...
+
+## Requirements
+
+### Storie
+
+As a Storage Administrator, I want audit logs that describe who perform an action on a S3 or IAM ressource, in order connect my Scality products to my IBM QRadar system.
+
+Acceptance Criteria:
+
+* Audit logs must be formatted using LEEF format
```
Comment (anurag4DSB) [inline docs/requirements/AUDIT-LOGS.md:15]: It would be nice to know which LEEF version are we going to transform the logs.
The headers might differ based on the version. The header will look something like this:
• LEEF:Version|Vendor|Product|Version|EventID|
• LEEF:1.0|Microsoft|MSExchange|4.0 SP1|15345|
• LEEF:2.0|Lancope|StealthWatch|1.0|41|^|

File: docs/requirements/AUDIT-LOGS.md [hunk 7]
```diff
@@ -0,0 +1,292 @@
...
+* Audit logs should be included into sreport
+* Audit logs rotation, purge, activation, desactivation must be configurable
+* Audit logs configuration, activation, desactivation must be documented
+* Audit logs configuration, activation, desactivation must persist to Scality upgrade
+
+### Who definition
+
+'Who' can be:
+
+* Account root user
```
Comment (anurag4DSB) [inline docs/requirements/AUDIT-LOGS.md:28]: If we to audit actions performed on an account, like create account of modify its entities, like quota, keys, injecting account IDs etc. Wo can be the master admin as well, whose credentials are used to create accounts.

File: docs/requirements/AUDIT-LOGS.md [hunk 8]
```diff
@@ -0,0 +1,292 @@
...
+* IAM users
+* Federated users (using web identity or SAML federation)
+
+### IAM ressources definition
+
+IAM ressources can be:
```
Comment (anurag4DSB) [inline docs/requirements/AUDIT-LOGS.md:34]: What about accounts?

Comment (nloewensen) [reply docs/requirements/AUDIT-LOGS.md:34]: I agreed that accounts need to be included, specifically, since we cannot ensure that the customer will not use his/her account-credentials to do some actions.

In addition, the action of the super-admin (account creation/deletion/modification) need to be logged as well

File: docs/requirements/AUDIT-LOGS.md (hunk 6 above)
Comment (nloewensen) [reply docs/requirements/AUDIT-LOGS.md:15]: Charles and team already have a good understanding of the formatting and its options, so please check with them for details.

File: docs/requirements/AUDIT-LOGS.md [hunk 9]
```diff
@@ -0,0 +1,292 @@
+# [S3C-2675](https://scality.atlassian.net/browse/S3C-2675): Audit logs requirements document
+
+## Context
+
+Some customers are using tools like [IBM QRadar](https://www.ibm.com/fr-fr/security/security-intelligence/qradar) in order to centralize and monitor who did what on IT systems for security reason. This tool is ingesting events formatted in Leefx format, which is a basic key/value list, separated with a space.
+
```
Comment (nloewensen) [inline docs/requirements/AUDIT-LOGS.md:6]: I am not clear if we will provide audit-logs in LEEF format "only" or if we plan to provide them by default in another format and only on customer request the LEEF format will be enalbed as as addition? This should be fully documented, so that ideally customers can enable it themselves.

Comment (nloewensen) [reply docs/requirements/AUDIT-LOGS.md:6]: One addition for Amadeus: they do not require the logs in "real-time" but poll the logs every few minutes and process them about every 10min.

File: docs/requirements/AUDIT-LOGS.md [hunk 10]
```diff
@@ -0,0 +1,292 @@
...
+As a Storage Administrator, I want audit logs that describe who perform an action on a S3 or IAM ressource, in order connect my Scality products to my IBM QRadar system.
+
+Acceptance Criteria:
+
+* Audit logs must be formatted using LEEF format
+* Audit logs must be self-descriptive, that means a Storage Administrator don't need to correlate multiple log lines or log files to know who perform an action on a ressource
+* In case of an integration with an External Identity Provider(using web identity or SAML federation), the 'who' must be the External Identity login or username
+* Failed action must appear in the Audit logs with the reason of failed (e.g: wrong signature, or privileges...)
+* Audit logs should be included into sreport
```
[the remark this answers is not in the document]
Comment (scality-trevorbenson) [reply docs/requirements/AUDIT-LOGS.md:19]: There is a project called [SOSCleaner](https://github.com/soscleaner/soscleaner) to obfuscate sensitive information from sosreports. It is available via EPEL.

File: docs/requirements/AUDIT-LOGS.md (hunk 6 above)
Comment (anurag4DSB) [reply docs/requirements/AUDIT-LOGS.md:15]: So shouldn't it be a part of requirements?

File: docs/requirements/AUDIT-LOGS.md (hunk 3 above)
Comment (RachedBenMustapha) [reply docs/requirements/AUDIT-LOGS.md:23]: I will depend entirely on the implementation whether it's easy to make immutable (flat files vs something else), and whether we can rely solely on containerization for isolation or if more mechanisms are needed.

I consider durability and resilience to be as important as immutability: if part of the audit trail can be lost or unavailable when part of the system is down (permanently or temporarily) say before it has been "published" to the IBM software, then there can be a time period during which there is no audit, which tend to be the exact time when the audit trail would need to be consulted.

File: docs/requirements/AUDIT-LOGS.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (scality-trevorbenson) [reply docs/requirements/AUDIT-LOGS.md:19]: > assuming Rached only refers to the "include into sreport". Indeed, when including them in sreport or sending them to our Support in any other way, we need to stripped/obfuscate/replace any sensitive information (AD-names, user/account names etc.) Some customers even consider IP-addresses as sensitive

@nloewensen There is a project called SOSCleaner to obfuscate sensitive information from sosreports. It is available via EPEL.

File: docs/requirements/AUDIT-LOGS.md [hunk 11]
```diff
@@ -0,0 +1,292 @@
+# [S3C-2675](https://scality.atlassian.net/browse/S3C-2675): Audit logs requirements document
+
+## Context
+
+Some customers are using tools like [IBM QRadar](https://www.ibm.com/fr-fr/security/security-intelligence/qradar) in order to centralize and monitor who did what on IT systems for security reason. This tool is ingesting events formatted in Leefx format, which is a basic key/value list, separated with a space.
+
+## Requirements
+
+### Storie
```
Comment (NicolasT) [inline docs/requirements/AUDIT-LOGS.md:9]:
```suggestion
### Story
```

File: docs/requirements/AUDIT-LOGS.md (hunk 5 above)
Comment (NicolasT) [inline docs/requirements/AUDIT-LOGS.md:5]: Is there some link to the spec of this *Leefx* format?

File: docs/requirements/AUDIT-LOGS.md (hunk 2 above)
Comment (NicolasT) [inline docs/requirements/AUDIT-LOGS.md:11]:
```suggestion
As a Storage Administrator, I want audit logs that describe who perform an action on a S3 or IAM ressource, in order to connect my Scality products to my IBM QRadar system.
```

Comment (NicolasT) [reply docs/requirements/AUDIT-LOGS.md:11]: FWIW, I asked the whole architect team to spend some time reading this spec, since its essence goes way beyond S3C (and future Zenko): we should figure out how to provide a unified audit logging system throughout our software stack (indeed likely not including OS-level auditing which is outside our software's scope), not end up with a different approach per component.

File: docs/requirements/AUDIT-LOGS.md (hunk 6 above)
Comment (NicolasT) [reply docs/requirements/AUDIT-LOGS.md:15]: Indeed, this ought to be defined. Furthermore, in the intro, there's a reference to *Leefx*. What's the right name?

File: docs/requirements/AUDIT-LOGS.md (hunk 4 above)
Comment (NicolasT) [inline docs/requirements/AUDIT-LOGS.md:16]:
```suggestion
* Audit logs must be self-descriptive, that means a Storage Administrator don't need to correlate multiple log lines or log files to know who perform an action on a resource
```

File: docs/requirements/AUDIT-LOGS.md (hunk 3 above)
Comment (NicolasT) [reply docs/requirements/AUDIT-LOGS.md:23]: Agreed. Also, knowing (part of) an audit trail is 'missing' could be an important requirement.

File: docs/requirements/AUDIT-LOGS.md (hunk 8 above)
Comment (NicolasT) [inline docs/requirements/AUDIT-LOGS.md:34]:
```suggestion
IAM resources can be:
```

File: docs/requirements/AUDIT-LOGS.md [hunk 12]
```diff
@@ -0,0 +1,292 @@
...
+
+IAM ressources can be:
+
+* Users
+* Policies
+* Groups
+* Identity provided by Identity provider
+* Roles
+
+## IAM Actions definition
```
Comment (NicolasT) [inline docs/requirements/AUDIT-LOGS.md:42]: Instead of semi-exhaustively listing these calls: are there any calls *not* listed here? If so, why? If not, shouldn't the simply requirement be "All operations are to be logged", given we may extend these APIs in the future anyway.

File: docs/requirements/AUDIT-LOGS.md [hunk 13]
```diff
@@ -0,0 +1,292 @@
...
+As a Storage Administrator, I want audit logs that describe who perform an action on a S3 or IAM ressource, in order connect my Scality products to my IBM QRadar system.
+
+Acceptance Criteria:
+
+* Audit logs must be formatted using LEEF format
+* Audit logs must be self-descriptive, that means a Storage Administrator don't need to correlate multiple log lines or log files to know who perform an action on a ressource
+* In case of an integration with an External Identity Provider(using web identity or SAML federation), the 'who' must be the External Identity login or username
+* Failed action must appear in the Audit logs with the reason of failed (e.g: wrong signature, or privileges...)
+* Audit logs should be included into sreport
+* Audit logs rotation, purge, activation, desactivation must be configurable
```
Comment (benoit-a) [inline docs/requirements/AUDIT-LOGS.md:20]: Does it mean "rotation = yes/no , purge = yes / no" or something more subtle like "max MB rotation / purge" or "max 3 days", etc... ?

File: docs/requirements/AUDIT-LOGS.md [hunk 14]
```diff
@@ -0,0 +1,292 @@
...
+* GetBucketTagging
+* GetBucketVersioning
+* GetBucketWebsite
+* GetObject
+* GetObjectAcl
+* GetObjectLegalHold
+* GetObjectLockConfiguration
+* GetObjectRetention
+* GetObjectTagging
+* GetObjectTorrent
```
Comment (benoit-a) [inline docs/requirements/AUDIT-LOGS.md:246]: Do we support that ?

File: docs/requirements/AUDIT-LOGS.md [hunk 15]
```diff
@@ -0,0 +1,292 @@
...
+* Objects
+
+### S3 actions definition
+
+S3 actions can be:
+
+* AbortMultipartUpload
+* CompleteMultipartUpload
+* CopyObject
+* CreateBucket
```
Comment (benoit-a) [inline docs/requirements/AUDIT-LOGS.md:203]: Question: do we plan to log any additional to the request ? like creating a Bucket "public" or something like that ?

Same for acl below. Logging "change of acl" is not very useful if you don't know anything that was changed. (assuming the purpose of the log is to trace someone's actions)

File: docs/requirements/AUDIT-LOGS.md (hunk 12 above)
Comment (benoit-a) [reply docs/requirements/AUDIT-LOGS.md:42]: I would agree to Nicolas comment.

Maybe that's just a wording but I would rephrase like:
"All IAM actions needs to be logged" (and listing exceptions if they are any) . Eventually, the list can be introduced with "including those actions (but not limited to) below"

On a moving target, a future proof definition (of what should be part of the log) should be included alongside a list.

File: docs/requirements/AUDIT-LOGS.md (hunk 2 above)
Comment (vrancurel) [reply docs/requirements/AUDIT-LOGS.md:11]: Can we just consider it is a log like other logs with a marker saying it is an audit log ? Then we don't have to redesign specialized distributed log system ...

Comment (NicolasT) [reply docs/requirements/AUDIT-LOGS.md:11]: What do you mean exactly with 'marker'? A specific field in the log output? Then how would this be handed over to an external system (assuming no post-processing)?

File: docs/requirements/AUDIT-LOGS.md [hunk 16]
```diff
@@ -0,0 +1,292 @@
...
+### Storie
+
+As a Storage Administrator, I want audit logs that describe who perform an action on a S3 or IAM ressource, in order connect my Scality products to my IBM QRadar system.
+
+Acceptance Criteria:
+
+* Audit logs must be formatted using LEEF format
+* Audit logs must be self-descriptive, that means a Storage Administrator don't need to correlate multiple log lines or log files to know who perform an action on a ressource
+* In case of an integration with an External Identity Provider(using web identity or SAML federation), the 'who' must be the External Identity login or username
+* Failed action must appear in the Audit logs with the reason of failed (e.g: wrong signature, or privileges...)
```
Comment (Stephane-Scality) [inline docs/requirements/AUDIT-LOGS.md:18]:
```suggestion
* Failed action must appear in the Audit logs with the reason of failure (e.g: wrong signature, or privileges...)
```

File: docs/requirements/AUDIT-LOGS.md (hunk 13 above)
Comment (Stephane-Scality) [reply docs/requirements/AUDIT-LOGS.md:20]: I believe rotation itself should provide guarantees, a user most likely is required to know that there are no holes in the log, so if rotation deletes a file it must be the oldest file on this system, failure to rotate/delete the oldest file stops rotation/deletion.

File: docs/requirements/AUDIT-LOGS.md (hunk 3 above)
Comment (Stephane-Scality) [reply docs/requirements/AUDIT-LOGS.md:23]: AFAIK most systems provide tampering detection rather than guaranteed tampering prevention (the guaranteed being very hard to do). It is in general sufficient to provide signatures that can be verified for a single trail, then for a group of trails, etc. in a "tree" form.
